### PR TITLE
feat: methylation candidates

### PR DESCRIPTION
**Pull request by Addimator**
### Description

It is now possible to find methylation candidates in a Fasta File. New subcommand is: --candidates  <input> [output]
               - input: FASTA File with the DNA of interest
               - output: Location of the resulting vcf file (optional, if not provided stdOut)

### QC

* [ ] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [ ] The documentation at https://github.com/varlociraptor/varlociraptor.github.io is updated in a separate PR to reflect the changes or this is not necessary (e.g. if the change does neither modify the calling grammar nor the behavior or functionalities of Varlociraptor).

### CODE REVIEW

**Comment by Addimator** [comment]
In order to run the test successfully, I have the required local file at test_meth_ev_1/genome.fasta. However, the current .gitignore configuration prevents the upload of this file, leading to test failures. I'm unsure whether adding an exception directly to the .gitignore is the right approach, considering there are no visible exceptions for files like candidates.matched.bcf.

File: src/cli.rs [hunk 1]
```diff
@@ -108,6 +108,23 @@ pub enum Varlociraptor {
         setting = structopt::clap::AppSettings::ColoredHelp,
     )]
     Genotype,
+    // TODO add subcommand for generating methylation candidates
+    #[structopt(
+        name = "candidates",
+        about = "Find methylation candidates for variants"
+    )]
+    Candidates {
```
Comment (johanneskoester) [inline src/cli.rs:116]:
```suggestion
    MethylationCandidates {
```

File: src/cli.rs [hunk 2]
```diff
@@ -108,6 +108,23 @@ pub enum Varlociraptor {
         setting = structopt::clap::AppSettings::ColoredHelp,
     )]
     Genotype,
+    // TODO add subcommand for generating methylation candidates
+    #[structopt(
+        name = "candidates",
```
Comment (johanneskoester) [inline src/cli.rs:113]:
```suggestion
        name = "methylation-candidates",
```

File: src/cli.rs [hunk 3]
```diff
@@ -108,6 +108,23 @@ pub enum Varlociraptor {
...
+    Candidates {
+        #[structopt(
+            name = "input",
+            parse(from_os_str),
+            required = true,
+            help = "Input FASTA File"
+        )]
+        input: PathBuf,
+
+        #[structopt(name = "output", parse(from_os_str), help = "Output VCF File")]
```
Comment (johanneskoester) [inline src/cli.rs:125]:
```suggestion
        #[structopt(name = "output", parse(from_os_str), help = "Output BCF File")]
```

File: src/candidates/methylation.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
+use bio::io::fasta;
+use rust_htslib::bcf::record::Numeric;
+use rust_htslib::bcf::{Format, Header, Writer};
+use std::collections::HashSet;
+use std::fs::File;
+use std::path::PathBuf;
```
Comment (johanneskoester) [inline src/candidates/methylation.rs:6]:
```suggestion
use std::path::PathBuf;
use anyhow::Result;
use itertools::Itertools;
```

File: src/candidates/methylation.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+/// Find all methylation candidates "CG" in a FASTA File
+///
+/// # Arguments
+///
+/// * `infasta` - path to FASTA with genome
+/// * `outbcf` - path to VCF with found methylation candidates (None for stdout)
+
+pub fn find_candidates(infasta: PathBuf, outvcf: Option<PathBuf>) {
+    // Open FASTA File
+    let fasta_file: File = File::open(infasta).expect("Unable to open");
```
Comment (johanneskoester) [inline src/candidates/methylation.rs:17]:
```suggestion
pub fn find_candidates(infasta: PathBuf, outvcf: Option<PathBuf>) -> Result<()> {
    // Open FASTA File
    let fasta_file: File = File::open(infasta).with_context(|| format!("Error reading FASTA file {infasta}"))?;
```

File: src/candidates/methylation.rs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            vcf_writer = Writer::from_path(path, &vcf_header, true, Format::Vcf).unwrap();
+        }
+        None => {
+            vcf_writer = Writer::from_stdout(&vcf_header, true, Format::Vcf).unwrap();
+        }
+    }
+
+    //Prepare the records
+    let mut record = vcf_writer.empty_record();
+    for rec in data {
```
Comment (johanneskoester) [inline src/candidates/methylation.rs:55]:
```suggestion
    for (contig, pos) in data {
```

File: src/candidates/methylation.rs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        record.set_alleles(new_alleles);
+        record.set_qual(f32::missing());
+
+        // Write record
+        match vcf_writer.write(&record) {
+            Ok(_) => {}
+            Err(e) => {
+                eprintln!("Error: {}", e)
+            }
+        }
```
Comment (johanneskoester) [inline src/candidates/methylation.rs:72]:
```suggestion
        vcf_writer.write(&record).with_context(|| format!("failed to write BCF record with methylation candidate")?;
```

File: src/candidates/methylation.rs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+use std::fs::File;
+use std::path::PathBuf;
+
+/// Find all methylation candidates "CG" in a FASTA File
+///
+/// # Arguments
+///
+/// * `infasta` - path to FASTA with genome
+/// * `outbcf` - path to VCF with found methylation candidates (None for stdout)
+
```
Comment (johanneskoester) [inline src/candidates/methylation.rs:14]:
```suggestion
```

File: src/candidates/methylation.rs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+pub fn find_candidates(infasta: PathBuf, outvcf: Option<PathBuf>) {
+    // Open FASTA File
+    let fasta_file: File = File::open(infasta).expect("Unable to open");
+    let reader = fasta::Reader::new(fasta_file);
+    let mut data: Vec<(String, i64)> = vec![];
+
+    // Collect all chromosomes and positions of candidates
+    for result in reader.records() {
```
Comment (johanneskoester) [inline src/candidates/methylation.rs:22]:
```suggestion
    // TODO: consider using an IndexedReader to write the BCF header first
    // and then write the methylation candidate records on the fly while reading
    // the FASTA.
    for result in reader.records() {
```

File: src/candidates/methylation.rs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+use std::path::PathBuf;
+
+/// Find all methylation candidates "CG" in a FASTA File
+///
+/// # Arguments
+///
+/// * `infasta` - path to FASTA with genome
+/// * `outbcf` - path to VCF with found methylation candidates (None for stdout)
+
+pub fn find_candidates(infasta: PathBuf, outvcf: Option<PathBuf>) {
```
Comment (johanneskoester) [inline src/candidates/methylation.rs:15]:
```suggestion
pub fn find_candidates(infasta: PathBuf, outvcf: Option<PathBuf>) {
    // TODO: add implementation for other methylation types (CHH, ..., given via a pattern arg)
```

**Review by johanneskoester — CHANGES_REQUESTED** [review]

File: src/candidates/methylation.rs [hunk 11]
```diff
@@ -0,0 +1,74 @@
+use anyhow::{Context, Ok, Result};
+use bio::io::fasta::Reader;
+use itertools::Itertools;
+use rust_htslib::bcf::record::Numeric;
+use rust_htslib::bcf::{Format, Header, Writer};
+use std::collections::HashSet;
+use std::fs::File;
+use std::path::PathBuf;
```
Comment (johanneskoester) [inline src/candidates/methylation.rs:8]: Above there are still some compiler warnings (unused imports)

**Review by johanneskoester — APPROVED** [review]

**Pull request merged**